### ISSUE TITLE
EVP_PKEY_CTX_get_ecdh_kdf_md doesn't work when using the FIPS provider

### ISSUE THREAD

**Issue by chrisccoulson** [openssl/openssl#19338]
If you run the following test case using both the default provider and the FIPS provider, you will notice that EVP_PKEY_CTX_get_ecdh_kdf_md incorrectly returns NULL when using the FIPS provider:
```
#include <openssl/core_names.h>
#include <openssl/decoder.h>
#include <openssl/ec.h>
#include <openssl/err.h>
#include <openssl/evp.h>
#include <openssl/params.h>
#include <openssl/provider.h>
#include <stdio.h>

static const unsigned char kExampleECKeyDER[] = {
    0x30, 0x77, 0x02, 0x01, 0x01, 0x04, 0x20, 0x71, 0x11, 0x11, 0x37, 0x68,
    0x6f, 0x3b, 0xe4, 0xc5, 0x20, 0x62, 0x77, 0xfc, 0x9e, 0x14, 0x35, 0x7a,
    0x90, 0x75, 0x2b, 0x84, 0x59, 0x14, 0xd6, 0x81, 0x3c, 0xfb, 0x97, 0x13,
    0x0f, 0x3e, 0x43, 0xa0, 0x0a, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d,
    0x03, 0x01, 0x07, 0xa1, 0x44, 0x03, 0x42, 0x00, 0x04, 0x01, 0x72, 0xdf,
    0x6b, 0x4c, 0xe2, 0x3a, 0x17, 0xd7, 0x6b, 0x41, 0xf7, 0x24, 0x33, 0xdd,
    0x62, 0xab, 0x6b, 0xd3, 0xaf, 0xbd, 0x1b, 0xdd, 0xc2, 0x5c, 0x94, 0x03,
    0xab, 0xe4, 0x05, 0x23, 0x38, 0x95, 0xc2, 0x33, 0x70, 0x35, 0xfe, 0xce,
    0x04, 0x3c, 0xc9, 0x06, 0x1d, 0x00, 0xba, 0x1f, 0x9b, 0xec, 0x95, 0x3c,
    0x63, 0x3d, 0x13, 0x53, 0x78, 0x6c, 0x14, 0x3e, 0xbd, 0x80, 0xdd, 0x98,
    0x51
};

int main(int argc, char **argv)
{
    const unsigned char *key = kExampleECKeyDER;
    size_t keylen = sizeof(kExampleECKeyDER);
    int libno = ERR_get_next_error_library();
    OSSL_DECODER_CTX *dctx = NULL;
    EVP_PKEY *pkey = NULL;
    EVP_PKEY_CTX *pctx = NULL;
    const EVP_MD *md;
    int ok = 0;

    if (argc != 2) {
        fprintf(stderr, "usage: %s <provider>\n", argv[0]);
        return 1;
    }

    if (OSSL_PROVIDER_load(NULL, argv[1]) == NULL) {
        fprintf(stderr, "cannot load %s provider\n", argv[1]);
        goto done;
    }
    if (OSSL_PROVIDER_load(NULL, "base") == NULL) {
        fprintf(stderr, "cannot load base provider\n");
        goto done;
    }

    if ((dctx = OSSL_DECODER_CTX_new_for_pkey(&pkey, NULL, NULL, NULL,
                                              OSSL_KEYMGMT_SELECT_ALL, NULL,
                                              NULL)) == NULL) {
        fprintf(stderr, "err: OSSL_DECODER_CTX_new_for_key\n");
        goto done;
    }
    if (!OSSL_DECODER_from_data(dctx, &key, &keylen)) {
        fprintf(stderr, "err: OSSL_DECODER_from_data\n");
        goto done;
    }
    if ((pctx = EVP_PKEY_CTX_new_from_pkey(NULL, pkey, NULL)) == NULL) {
        fprintf(stderr, "err: EVP_PKEY_CTX_new_from_pkey\n");
        goto done;
    }
    if (!EVP_PKEY_derive_init(pctx)) {
        fprintf(stderr, "err: EVP_PKEY_derive_init\n");
        goto done;
    }
    if (!EVP_PKEY_CTX_set_ecdh_kdf_md(pctx, EVP_sha256())) {
        fprintf(stderr, "err: EVP_PKEY_CTX_set_ecdh_kdf_md\n");
        goto done;
    }
    if (!EVP_PKEY_CTX_get_ecdh_kdf_md(pctx, &md)) {
        fprintf(stderr, "err: EVP_PKEY_CTX_get_ecdh_kdf_md\n");
        goto done;
    }
    if (md != EVP_sha256())
        fprintf(stderr, "unexpected md (%p)\n", md);
    else
        printf("success!\n");

    ok = 1;

 done:
    EVP_PKEY_CTX_free(pctx); 
    OSSL_DECODER_CTX_free(dctx);
    EVP_PKEY_free(pkey);

    if (!ok)
        ERR_print_errors_fp(stderr);

    return ok ? 0 : 1;
}
```

The output gives:
```
$ ./test default
success!
$ ./test fips
unexpected md ((nil))
```

The issue is that when `evp_pkey_ctx_ctrl_to_param` queries the digest from the ecdh EVP_KEYEXCH implementation that is on the FIPS provider side, the digest name that is returned is "SHA2-256" - this is the name that the EVP_MD is initialized with because it's the first name of the underlying algorithm. `evp_pkey_ctx_ctrl_to_param` then tries to map this name to a legacy EVP_MD in `fix_md`, using `evp_get_digestbyname_ex`. Because `evp_get_digestbyname_ex` looks up the names on the application's library context, the lookup fails because the application's library context hasn't loaded any digests yet. The application's library context only contains a mapping with the name "SHA256", which was created during the call to `OPENSSL_init_crypto`.

This has implications at least for `openssl-cms`, which currently ignores the `-keyopt ecdh_kdf_md:sha256` argument when using the FIPS provider and just uses SHA1 for the key agreement, because of this code here: https://github.com/openssl/openssl/blob/8c7d847e2e6ac6bfded210c19fd8461254bb2be3/crypto/cms/cms_ec.c#L309

(Tested with 8c7d847e2e6ac6bfded210c19fd8461254bb2be3)

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of the 3.4 dev cycle barring further input

**Comment by t8m** [comment]
This is very interesting corner case bug that we should IMO investigate fix for.

**Comment by 0140454** [comment]
It seems that my report #24004 is duplicated with this one.

Should I close mine?

**Comment by slontis** [comment]
Changing the above code to   
```` 
    cmd = EVP_MD_fetch(NULL, "SHA256", NULL);
    if (!EVP_PKEY_CTX_set_ecdh_kdf_md(pctx, cmd)) {
        fprintf(stderr, "err: EVP_PKEY_CTX_set_ecdh_kdf_md\n");
        goto done;
    }
    if (!EVP_PKEY_CTX_get_ecdh_kdf_md(pctx, &md)) {
        fprintf(stderr, "err: EVP_PKEY_CTX_get_ecdh_kdf_md\n");
        goto done;
    }
````
fixes the failure

**Comment by slontis** [comment]
The call to EVP_PKEY_CTX_set_ecdh_kdf_md(pctx, EVP_sha256())) works correctly (i.e. it does a fetch inside ecdh_ecxh.c)
So it is the getter that the issue (especially if it has code in cms that sets the value to SHA1 if the get returns NIL)

**Comment by slontis** [comment]
It doesnt matter what I pass to the set..

calling EVP_MD_fetch(NULL, "SHA256", NULL); anywhere populates the namemap otherwise it doesnt.

The name returned from the ecdh_exch is "SHA2-256" for both cases

fix_cipher_md() calls get_algo_by_name() which calls get_md_by_name() which calls evp_get_digestbyname_ex().

````
    dp = (const EVP_MD *)OBJ_NAME_get(name, OBJ_NAME_TYPE_MD_METH);

    if (dp != NULL)
        return dp;

    namemap = ossl_namemap_stored(libctx);
    id = ossl_namemap_name2num(namemap, name);
    if (id == 0)
        return NULL;
````

dp is NULL since there is no alias for "SHA2-256"
And without the call to EVP_MD_fetch() id returns 0.

**Comment by slontis** [comment]
@t8m @nhorman @levitte any thought on this?

**Comment by levitte** [comment]
Yeah, our "on-demand" fetching has consequences.  Like @slontis says, a workaround is indeed to fetch something known (preferably from the intended provider, see suitable properties) to get the proverbial ball rolling...

**Comment by t8m** [comment]
Fix in #24940
